### ISSUE TITLE
False positive `undefined-variable` for variable in inner scope matching outer scope type annotation

### ISSUE THREAD

**Issue by jacobtylerwalls** [PyCQA/pylint#5326]
### Bug description

```python
def func():
    a: bool
    _ = [print(kwarg1=a, kwarg2=a) for a in range(10)]
```

### Command used

```shell
pylint a.py
```

### Pylint output

```shell
a.py:3: [E0602(undefined-variable), func] Undefined variable 'a'
```

### Expected behavior

No messages

### Pylint version

```shell
pylint 2.11.2-dev0
astroid 2.8.5
Python 3.9.6 (v3.9.6:db3ff76da1, Jun 28 2021, 11:49:53)

Does NOT occur on 2.11.1, possible regression in dev version?
```

**Comment by jacobtylerwalls** [comment]
After bisecting, the change occurred in b1c4735d2897c55e16c5c14b3926ba4b676df8c7. Reading that change, it seems `undefined-variable` is the new intended behavior, but it should be raised on line 2, not line 3. And if only one kwarg involves the variable, then `unused-variable` is emitted instead, which I suppose is a false negative for `undefined-variable`.

**Comment by IceTDrinker** [comment]
hello to pop in and say we have that issue but with the annotated outer scope variable name being **used**/assigned to later

Basically comething like:

```python
node: MyNode  # triggers the undefined-variable which is hard to comprehend
list_of_nodes = [node for node in maybe_node_list if isinstance(node, MyNode)

for node in other_list_of_node:
    pass
```

**Comment by IceTDrinker** [comment]
This also fails on the first line

```python
list_of_nodes = [node for node in maybe_node_list if isinstance(node, MyNode)]

node: MyNode

for node in other_list_of_node:
    pass
```

**Comment by jacobtylerwalls** [comment]
I broke the false negative out into #5624. That means this ticket is now just a false positive with a PR to close it.
